File: src/main/scala/resonantinduction/core/prefab/node/NodePressure.java
package resonantinduction.core.prefab.node;

import net.minecraftforge.common.util.ForgeDirection;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.IFluidHandler;
import universalelectricity.api.core.grid.INodeProvider;
import universalelectricity.api.core.grid.IUpdate;

import java.util.Map;

/**
 * Created by robert on 8/15/2014.
 */
public class NodePressure extends NodeTank implements IUpdate
{
    private int pressure = 0;

    public NodePressure(INodeProvider parent)
    {
        super(parent);
    }

    public NodePressure(INodeProvider parent, int buckets)
    {
        super(parent, buckets);
    }

    @Override
    public void update(double deltaTime)
    {
        if (!world().isRemote)
        {
            updatePressure();
            if(getFluid() != null) {
                for (Map.Entry<Object, ForgeDirection> entry : connections.entrySet()) {
                    if (entry.getKey() instanceof INodeProvider && ((INodeProvider) entry.getKey()).getNode(NodePressure.class, entry.getValue().getOpposite()) instanceof NodePressure) {
                        NodePressure node = (NodePressure) ((INodeProvider) entry.getKey()).getNode(NodePressure.class, entry.getValue().getOpposite());
                        if (node.getPressure(entry.getValue().getOpposite()) <= getPressure(entry.getValue())) {

                        }
                    } else if (entry.getKey() instanceof INodeProvider && ((INodeProvider) entry.getKey()).getNode(NodeTank.class, entry.getValue().getOpposite()) instanceof NodeTank) {
                        NodeTank node = (NodeTank) ((INodeProvider) entry.getKey()).getNode(NodeTank.class, entry.getValue().getOpposite());
                        if (node.canFill(entry.getValue().getOpposite(), getFluid().getFluid()))
                        {
                            FluidStack stack = drain(Integer.MAX_VALUE, false);
                            int drained = node.fill(stack, true);
                            drain(drained, true);
                        }
                    } else if (entry.getKey() instanceof IFluidHandler)
                    {
                        if(((IFluidHandler) entry.getKey()).canFill(entry.getValue().getOpposite(), getFluid().getFluid()))
                        {
                            FluidStack stack = drain(Integer.MAX_VALUE, false);
                            int drained = ((IFluidHandler) entry.getKey()).fill(entry.getValue().getOpposite(), stack, true);
                            drain(drained, true);
                        }
                    }
                }
            }
        }
    }

    @Override
    public boolean canUpdate()
    {
        return true;
    }

    @Override
    public boolean continueUpdate()
    {
        return true;
    }

    protected void updatePressure()
    {
        int totalPressure = 0;
        int connectionSize = connections.size();
        int minPressure = 0;
        int maxPressure = 0;

        for(Map.Entry<Object, ForgeDirection> entry : connections.entrySet())
        {
            if(entry.getKey() instanceof INodeProvider &&  ((INodeProvider) entry.getKey()).getNode(NodePressure.class, entry.getValue().getOpposite()) instanceof NodePressure)
            {
                NodePressure node = (NodePressure) ((INodeProvider) entry.getKey()).getNode(NodePressure.class, entry.getValue().getOpposite());
                int pressure = node.getPressure(entry.getValue().getOpposite());
                minPressure = Math.min(pressure, minPressure);
                maxPressure = Math.max(pressure, maxPressure);
                totalPressure += pressure;
            }
        }

        if (connectionSize == 0)
        {
            setPressure(0);
        }
        else
        {
            if (minPressure < 0)
            {
                minPressure += 1;
            }
            if (maxPressure > 0)
            {
                maxPressure -= 1;
            }

            setPressure(Math.max(minPressure, Math.min(maxPressure, totalPressure / connectionSize + Integer.signum(totalPressure))));
        }
    }

    public int getPressure(ForgeDirection direction) {
        return pressure;
    }

    public void setPressure(int pressure) {
        this.pressure = pressure;
    }
}

File: src/main/scala/resonantinduction/core/prefab/node/NodePressure.scala
package resonantinduction.core.prefab.node

import net.minecraftforge.common.util.ForgeDirection
import net.minecraftforge.fluids.{FluidContainerRegistry, IFluidHandler}
import resonant.api.grid.{INodeProvider, IUpdate}
import resonant.lib.grid.UpdateTicker
import resonant.lib.prefab.fluid.NodeFluid

import scala.collection.convert.wrapAll._

/**
 * A node for fluid that moves based on pressure
 *
 * @param parent Parent(TileEntity or Multipart) that contains this node
 * @param volume Amount of fluid in liters
 *
 * @author Calclavia
 */
class NodePressure(parent: INodeProvider, volume: Int = FluidContainerRegistry.BUCKET_VOLUME) extends NodeFluid(parent, volume) with IUpdate
{
  var maxFlowRate = 20
  var maxPressure = 100
  private var _pressure: Int = 0

  UpdateTicker.addUpdater(this)

  def update(deltaTime: Double)
  {
    println(world)
    if (!world.isRemote)
    {
      updatePressure()
      distribute(deltaTime)
    }
  }

  def distribute(deltaTime: Double)
  {
    val flowRate = (maxFlowRate * deltaTime).toInt

    directionMap.foreach
    {
      case (handler: IFluidHandler, dir: ForgeDirection) =>
      {
        if (handler.isInstanceOf[NodePressure])
        {
          //"A" is this node. "B" is the other node
          //It's another pressure node
          val otherNode = handler.asInstanceOf[NodePressure]
          val pressureA = pressure(dir)
          val pressureB = otherNode.pressure(dir.getOpposite)

          //High pressure to low
          if (pressureA >= pressureB)
          {
            val tankA = getPrimaryTank

            if (tankA != null)
            {
              val fluidA = tankA.getFluid

              if (fluidA != null)
              {
                val amountA = fluidA.amount

                if (amountA > 0)
                {
                  val tankB = otherNode.getPrimaryTank

                  if (tankB != null)
                  {
                    doDistribute(dir, this, otherNode, flowRate)
                  }
                }
              }
            }
          }
        }
        else
        {
          //It's a fluid handler.
          val pressure = this.pressure(dir)
          val tankPressure = 0
          val sourceTank = getPrimaryTank
          val transferAmount = (Math.max(pressure, tankPressure) - Math.min(pressure, tankPressure)) * flowRate

          if (pressure > tankPressure)
          {
            if (sourceTank.getFluidAmount > 0 && transferAmount > 0)
            {
              val drainStack = drain(dir.getOpposite, transferAmount, false)
              drain(dir.getOpposite, handler.fill(dir.getOpposite, drainStack, true), true)
            }
          }
          else if (pressure < tankPressure)
          {
            if (transferAmount > 0)
            {
              val drainStack = handler.drain(dir.getOpposite, transferAmount, false)
              if (drainStack != null)
              {
                handler.drain(dir.getOpposite, fill(dir.getOpposite, drainStack, true), true)
              }
            }
          }
        }
      }
    }
  }

  protected def doDistribute(dir: ForgeDirection, nodeA: NodePressure, nodeB: NodePressure, flowRate: Int)
  {
    val tankA = nodeA.getPrimaryTank
    val tankB = nodeB.getPrimaryTank
    val pressureA = nodeA.pressure(dir)
    val pressureB = nodeB.pressure(dir.getOpposite)
    val amountA = tankA.getFluidAmount
    val amountB = tankB.getFluidAmount

    var quantity = Math.max(if (pressureA > pressureB) (pressureA - pressureB) * flowRate else Math.min((amountA - amountB) / 2, flowRate), Math.min((amountA - amountB) / 2, flowRate))
    quantity = Math.min(Math.min(quantity, tankB.getCapacity - amountB), amountA)

    if (quantity > 0)
    {
      val drainStack = drain(dir.getOpposite, quantity, false)
      if (drainStack != null && drainStack.amount > 0)
      {
        drain(dir.getOpposite, nodeB.fill(dir, drainStack, true), true)
      }
    }
  }

  protected def updatePressure()
  {
    var totalPressure = 0
    val connectionSize = connections.size
    var minPressure = 0
    var maxPressure = 0

    directionMap.foreach
    {
      case (handler: IFluidHandler, dir: ForgeDirection) =>
      {
        if (handler.isInstanceOf[NodePressure])
        {
          val node = handler.asInstanceOf[NodePressure]
          val pressure = node.pressure(dir.getOpposite)
          minPressure = Math.min(pressure, minPressure)
          maxPressure = Math.max(pressure, maxPressure)
          totalPressure += pressure
        }
      }
    }

    if (connectionSize == 0)
    {
      pressure = 0
    }
    else
    {
      if (minPressure < 0)
        minPressure += 1
      if (maxPressure > 0)
        maxPressure -= 1

      pressure = Math.max(minPressure, Math.min(maxPressure, totalPressure / connectionSize + Integer.signum(totalPressure)))
    }
  }

  def pressure: Int = _pressure

  def pressure(direction: ForgeDirection): Int = _pressure

  def pressure_=(pressure: Int)
  {
    this._pressure = pressure
  }

  def canUpdate = !isInvalid

  def continueUpdate = !isInvalid
}